examples/piggybank/src/Root.tsx

```diff
@@ -43,14 +44,39 @@ const isPiggybankSmashed = (state: PiggyBankState): state is PiggyBankStateSmash
 
 const client = new JsonRpcClient(new HttpProvider(JSON_RPC_URL));
 
-/** Promise resolves when callback is called from the extension, letting us know that the wallet API is ready for use. */
-const apiReady = new Promise<void>((resolve) => {
-    if (window.concordium !== undefined) {
-        resolve();
-    } else {
-        window.concordiumReady = resolve;
-    }
-});
+/**
+ * Promise resolves to the Concordium provider when it has been successfully injected into
+ * the window and is ready for use.
+ * @param timeout determines how long to wait before rejecting if the Concordium provider is not available, in milliseconds.
+ * @returns a promise containing the Concordium wallet provider API.
+ */
+// TODO This function should be made available from the web-sdk for ease of use.
+async function detectConcordiumProvider(timeout = 5000): Promise<WalletApi> {
+    return new Promise((resolve, reject) => {
+        let alreadyHandled = false;
+
+        function handleConcordium() {
+            if (alreadyHandled) {
+                return;
+            }
+            alreadyHandled = true;
+
+            if (window.concordium) {
+                resolve(window.concordium);
+            }
+            reject();
+        }
+
+        if (window.concordium) {
+            handleConcordium();
+        } else {
+            window.addEventListener('concordium#initialized', handleConcordium, { once: true });
+            setTimeout(() => {
+                handleConcordium();
+            }, timeout);
+        }
```

Comment:
All the `alreadyHandled` stuff shouldn't be necessary as a promise only resolves once. Can be simplified by something like:

```suggestion
        if (window.concordium) {
            resolve(window.concordium);
        } else {
            const t = setTimeout(reject, timeout);
            window.addEventListener('concordium#initialized', () => {
                resolve(window.concordium as WalletApi);
                clearTimeout(t);
            }, { once: true });
        }
```

Probably needs to be implemented locally to not annoy the linter too much.

Reply:
I ended up with a mixture of what I had and your suggestion that makes it a bit nicer.